### ISSUE TITLE
mingw support

### ISSUE THREAD

**Issue by gwald** [horde3d/Horde3D#206]
Originally raised here: http://horde3d.org/forums/viewtopic.php?f=2&t=2326

I know mingw isn't officially supported, but it's a great alternative to VS and seems to be very close to working!

It builds both GLFW and SDL2 and libHorde3D, but gets stuck at linking the libSampleFramework: 
undefined reference to `__imp_h3dutLoadResourcesFromDisk'

Suspect it's a cmake make issue.

**Comment by gwald** [comment]
looks like it should just work, there a lot of __MINGW32__ which should also work with x64

Using dependency walker (dll viewer) the function is there
libHorde3DUtils.dll.a is 9k
the obj and objects.a is +300kb
I tried creating the dll using the object.a, but still get the same dll size 2.7mb

I don't know how to debug this issue :/

**Comment by algts** [comment]
Hello.
Current findings: generator does not matter (either it is ninja or msys makefiles) as well as compiler (tried with msys clang).
There are several points to consider:
- link order is wrong (gcc has problems with it, clang seem to be using mingw linker, so it can have the same problems)
- something has to be changed in cmake files (additional flags for mingw or something like that).
- bugs in mingw linker (unlikely).

Objdump shows that the function is there (h3dutLoadResourcesFromDisk), but it seems that there are no __imp functions in the generated dll. So it can be a c++ mangling issue.

**Comment by gwald** [comment]
it's very strange, there are no horde3d  __imp pointers in libhorde3d.dll and the libhorde3dutils.dll  has some that look like it should be in libhorde3d.dll
```
0000000068926338 I __imp_h3dAddResource
0000000068926340 I __imp_h3dCastRay
0000000068926348 I __imp_h3dGetCameraProjMat
0000000068926350 I __imp_h3dGetCastRayResult
0000000068926358 I __imp_h3dGetMessage
0000000068926360 I __imp_h3dGetNodeParamF
0000000068926368 I __imp_h3dGetNodeParamI
0000000068926370 I __imp_h3dGetNodeTransMats
0000000068926378 I __imp_h3dGetRenderTargetData
0000000068926380 I __imp_h3dGetResName
0000000068926388 I __imp_h3dGetResType
0000000068926390 I __imp_h3dGetVersionString
0000000068926398 I __imp_h3dLoadResource
00000000689263a0 I __imp_h3dQueryUnloadedResource
```

**Comment by gwald** [comment]
I'm not sure why, but the dll's work when building without the examples.
 cmake  -G "Unix Makefiles" -DHORDE3D_BUILD_EXAMPLES=OFF ..

Using those dll's to overwrite the dll.a in engine/utils folders and rerunning the make, continues to build the examples correctly.

Sorry I don't know how to fix what's happening, but at least this is an easy work around! 
I will also test in linux and with sdl2 later, I except it to also work
:)

**Comment by algts** [comment]
Hello. I've found the solution - you have to change library order from this:
`target_link_libraries(Chicago Horde3D Horde3DUtils SampleFramework ${GLFW_LIBRARIES})`
to this:
`target_link_libraries(Chicago SampleFramework Horde3D Horde3DUtils ${GLFW_LIBRARIES})`

Then everything compiles correctly.
I should now check whether it affects other compilers. If you have time please also check on Linux, does it have any side effects? If not, then this will be the main compiling route, otherwise additional compiler check should be done in cmake files.

**Comment by gwald** [comment]
I made the changes above to the samples and terrain extension to a fresh master, all my systems work!
Tested
native mint linux 20.3 GCC 9.4.0
win7 mingw GCC 10.2.0
linux cross compile via mingw 10.0.0

Tested with GLFW and all work as expected!
Thanks man!! :) 
:+1:

**Comment by algts** [comment]
Fixed in latest master and release 2.0.2. Closing.
